parquet-column/src/main/java/org/apache/parquet/schema/Types.java

```diff
@@ -555,114 +622,1267 @@ protected GroupType build(String name) {
+
+      public MapListValueBuilder<P, Q> requiredListValue() {
+        return listValue(Type.Repetition.REQUIRED);
+      }
+
+      public MapListValueBuilder<P, Q> optionalListValue() {
+        return listValue(Type.Repetition.OPTIONAL);
+      }
+
+      public BaseMapBuilder<P, Q> value (Type type) {
+        parent.setValueType(type);
+        return this.parent;
+      }
+
+      public BaseMapBuilder<P, Q> requiredValue (Type type) {
+        return value(type);
+      }
+
+      public BaseMapBuilder<P, Q> optionalValue (Type type) {
+        return value(type);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseGroupBuilder<P, MapGroupValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapGroupValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapMapValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseMapBuilder<P, MapMapValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapMapValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapMapValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapListValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListKeyBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListKeyBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        return super.named(name);
+      }
+
+      @Override
+      protected MapListKeyBuilder<P, Q> self() {
+        return this;
+      }
+
+      public void setKey(Type key) {
+        parent.setKeyType(key);
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
```

Comment:
As these are not public methods, I did not want to put any contracts. However, it definitely won't hurt. Adding the checks.